Refactor creation of aggregated data + regional chloropleth

lizgzil opened the pull request:
---

# Description

This looks big, but do not fear, it is mostly a refactor of the notebook we both created in https://github.com/nestauk/dap_prinz_green_jobs/pull/88
- `dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py` is all the data processing from the existing notebook
- Script to create the per occupation aggregated dataset (as before)
- Script to create the per region (ITL3) aggregated dataset
- [Notebook](dap_prinz_green_jobs/notebooks/regional_comparison.ipynb) to create a chloropleth of green measures per region
- [Notebook](dap_prinz_green_jobs/notebooks/regional_comparison_by_occ.py) to create a chloropleth of green measures per region PER OCC
- Chloropleth utils (getting the geometries per itl)

![Screenshot 2023-11-10 at 18 27 35](https://github.com/nestauk/dap_prinz_green_jobs/assets/15956065/84d6d944-e35d-46a3-b2ed-a373c0f97064)

Per SOC:
![Screenshot 2023-12-06 at 09 47 29](https://github.com/nestauk/dap_prinz_green_jobs/assets/15956065/e322849b-faa1-4db4-a2f1-7c9de9daf33c)


Fixes #93 and helps with #91 

# Instructions for Reviewer
Try running:

```
python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc.py

```
and try to run the code in `dap_prinz_green_jobs/notebooks/regional_comparison.ipynb` to see if the chlorpleth is generated ok. 

(it is also saved here https://drive.google.com/file/d/1_6vkraQH_UDM5L_WHYnJcqBThiUclcui/view)

Is there any better way to plot this?

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

india-kerle commented on dap_prinz_green_jobs/analysis/README.md:
````diff
@@ -0,0 +1,21 @@
+Create a dataset of green measures aggregated by occupation by running
+
+```
+python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc.py
````
this ran for me no problems and created that csv!

india-kerle commented on dap_prinz_green_jobs/analysis/README.md:
````diff
@@ -0,0 +1,21 @@
+Create a dataset of green measures aggregated by region (ITL3) by running
+
+```
+python dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py
````
also ran with no errors

india-kerle commented on dap_prinz_green_jobs/analysis/README.md:
````diff
@@ -0,0 +1,21 @@
+
+```
+
+This will create a csv `outputs/data/ojo_application/extracted_green_measures/analysis/occupation_aggregated_data_{DATE}.csv`, where each row is aggregated green measures information for each occupation.
````
It saved 4 datasets for me to s3

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+from dap_prinz_green_jobs.analysis.ojo_analysis.process_ojo_green_measures import *
```
```suggestion
import dap_prinz_green_jobs.analysis.ojo_analysis.process_ojo_green_measures as pg
```

india-kerle replied:
Could be good to add a quick one liner about what this script does and how to run it

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+        occs_measures_df,
+        inds_measures_df,
+        soc_name_dict,
+    ) = load_ojo_green_measures(
+        production, config, skills_date_stamp, occ_date_stamp, ind_date_stamp
+    )
```
```suggestion
    ) = pg.load_ojo_green_measures(
        production, config, skills_date_stamp, occ_date_stamp, ind_date_stamp
    )
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+        production, config, skills_date_stamp, occ_date_stamp, ind_date_stamp
+    )
+
+    all_green_measures_df = merge_green_measures(
+        skill_measures_df, occs_measures_df, inds_measures_df, soc_name_dict
+    )
```
```suggestion
    all_green_measures_df = pg.merge_green_measures(
        skill_measures_df, occs_measures_df, inds_measures_df, soc_name_dict
    )
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+
+    # Add some additional data
+
+    all_green_measures_df = add_salaries(all_green_measures_df, job_id_col=job_id_col)
+    all_green_measures_df = add_locations(all_green_measures_df, job_id_col=job_id_col)
+    all_green_measures_df = add_sic_info(all_green_measures_df)
```
```suggestion
    all_green_measures_df = pg.add_salaries(all_green_measures_df, job_id_col=job_id_col)
    all_green_measures_df = pg.add_locations(all_green_measures_df, job_id_col=job_id_col)
    all_green_measures_df = pg.add_sic_info(all_green_measures_df)
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+    all_green_measures_df = add_locations(all_green_measures_df, job_id_col=job_id_col)
+    all_green_measures_df = add_sic_info(all_green_measures_df)
+
+    all_skills_df = create_skill_df(skill_measures_df)
+
```
```suggestion
    all_skills_df = pg.create_skill_df(skill_measures_df)

```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+
+    agg_itl_by = "itl_3_code"  # itl_2_code or itl_3_code
+
+    itl_aggregated_data = create_agg_data(
+        all_green_measures_df,
+        all_skills_df,
+        soc_descriptions_dict=None,
+        agg_col=agg_itl_by,
+        job_id_col="job_id",
+    )
```
```suggestion
    itl_aggregated_data = pg.create_agg_data(
        all_green_measures_df,
        all_skills_df,
        soc_descriptions_dict=None,
        agg_col=agg_itl_by,
        job_id_col="job_id",
    )
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+    itl_aggregated_data["occ_greenness"] = itl_aggregated_data[
+        "average_occ_green_timeshare"
+    ].apply(
+        lambda x: categorical_assign(
```
```suggestion
        lambda x: pg.categorical_assign(
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+    itl_aggregated_data["ind_greenness"] = itl_aggregated_data[
+        "average_ind_perunit_ghg"
+    ].apply(
+        lambda x: categorical_assign(
```
```suggestion
        lambda x: pg.categorical_assign(
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+    itl_aggregated_data["skills_greenness"] = itl_aggregated_data[
+        "average_prop_green_skills"
+    ].apply(
+        lambda x: categorical_assign(
```
```suggestion
        lambda x: pg.categorical_assign(
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+    )
+
+    itl_aggregated_data["greenness_score"] = itl_aggregated_data.apply(
+        lambda x: get_one_score(
```
```suggestion
        lambda x: pg.get_one_score(
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc.py:
```diff
@@ -0,0 +1,82 @@
+from dap_prinz_green_jobs.analysis.ojo_analysis.process_ojo_green_measures import *
```
ditto for all the functions imported this way 

```suggestion
import dap_prinz_green_jobs.analysis.ojo_analysis.process_ojo_green_measures as pg
```

india-kerle replied:
ditto adding a comment at the beginning to describe the script and how to run it

lizgzil replied:
good point!

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_soc.py:
```diff
@@ -0,0 +1,82 @@
+if __name__ == "__main__":
+    # date stamps as defined in https://github.com/nestauk/dap_prinz_green_jobs/issues/75
+
+    production = "True"
+    config = "base"
+    skills_date_stamp = "20230914"
+    occ_date_stamp = "20231002"
+    ind_date_stamp = "20231013"
```
thinking about this - should we add this to an analysis config so we just need to update the dates in one place?

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+from tqdm import tqdm
+
+import os
+
```
from typing import List, Tuple, Dict, Union

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+
+
+# clean up skills
+def merge_ents(ents):
```
+ add doc strings
```suggestion
def merge_ents(ents: List[str]) -> List[str]:
```

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+
+
+# function to clean SOC names
+def clean_soc_name(soc_name):
```
docstrings + typing hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+        return None
+
+
+def load_ojo_green_measures(
+    production, config, skills_date_stamp, occ_date_stamp, ind_date_stamp
+):
```
doc strings + typing hints - should we add this in the ojo getter instead? up for debate!

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return skill_measures_df, occs_measures_df, inds_measures_df, soc_name_dict
+
+
+def merge_green_measures(
+    skill_measures_df,
+    occs_measures_df,
+    inds_measures_df,
+    soc_name_dict,
+    job_id_col="job_id",
+):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+        all_green_measures_df[soc_columns] = all_green_measures_df["SOC"].apply(
+            lambda x: x[soc_columns] if x else None
+        )
+    all_green_measures_df.drop(columns=["SOC"], inplace=True)
+
+    all_green_measures_df.rename(
+        columns={"name": "SOC_names", job_id_col: "job_id"}, inplace=True
+    )
```
could pipe these together i.e. (sudo-code!!)

```
all_green_measures_df = (all_green_measures_df
. drop()
.rename()
```
but its a nice to have

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    # ENTS: [[['research'], 'SKILL'], [['This is a cut up', 'a cut up sentence'], 'MULTISKILL']]
+    # to [[['research'], 'SKILL'], [['This is a cut up'], 'SKILL'], [['a cut up sentence'], 'SKILL']]
+
+    def separate_multiskill_ents(entlist):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return all_green_measures_df
+
+
+def add_salaries(all_green_measures_df, job_id_col="job_id"):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+
+
+def add_salaries(all_green_measures_df, job_id_col="job_id"):
+    salary_information = get_mixed_ojo_salaries_sample()
```
we're going to need to remember to update these getters when we use the large sample

lizgzil replied:
omg very good shout! I'm going to add an argument from a config to help with this

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return all_green_measures_df
+
+
+def add_locations(all_green_measures_df, job_id_col="job_id"):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return all_green_measures_df
+
+
+def add_sic_info(all_green_measures_df):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+
+def get_soc_info():
+    # we need soc descriptions as well - they have descriptions for 6 digit sic codes
+    soc_descriptions = pd.read_excel(
+        "s3://prinz-green-jobs/inputs/data/occupation_data/ons/extendedsoc2020structureanddescriptionsexcel121023.xlsx",
+        sheet_name="Extended SOC descriptions MG1-9",
+    )
```
is there a getter for this? I might be making that up though

lizgzil replied:
no - not with the descriptions - I'll add it though since this would be better

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return soc_descriptions_dict
+
+
+def add_green_topics(all_green_measures_df):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+        .dropna(subset=["subunit_group", "group_title", "descriptions"])
+    )
+    # clean description
+    soc_descriptions["clean_description"] = soc_descriptions.descriptions.apply(
+        lambda x: x.replace("\n", " ")
+        .replace("\r", " ")
+        .replace("\t", " ")
+        .replace("  ", " ")
+        .strip()
+        .replace("\xa0", "")
+    )
```
could add this to the text cleaning utils instead

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return all_green_measures_df
+
+
+def add_additional_metadata(all_green_measures_df, job_id_col="job_id"):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return all_green_measures_df, soc_descriptions_dict
+
+
+def filter_large_occs(
+    all_green_measures_df, min_num_job_ads=50, occ_col="SOC_2020_name"
+):
```
docstrings + type hints

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    return filtered_df
+
+
+def create_green_skills_df(all_green_measures_df, occ_col="SOC_2020_name"):
```
docstrings + type hints

india-kerle replied:
i've realised there are so many functions actually - i don't know how important docstrings + type hints are at this point given the analysis is very live and updating - maybe we just make sure to save a few days to clean up the analysis code once we're happy with the graphs

lizgzil replied:
I've added the type hints throughout. but I think docstrings could just be left for the most important functions rather than all of them. I agree, it feels a bit in flux to spend too much time on.

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    )
+    ents_explode = ents_explode.explode("skill_label").reset_index(drop=True)
+
+    skill_match_thresh = 0.7
```
could be nice to pull this out as a parameter if we want to play around with this later downstream

lizgzil replied:
yep good idea - I've added to a config

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+    job_id_col="job_id",
+):
+    """
+    Much like create_agg_occ_measures but more generic to aggregate by any column
```
nice !

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+                .value_counts()[0:5]
+                .to_dict()
+            )
+            top_5_not_green_skills_prop = (
+                filtered_skills[pd.isnull(filtered_skills["extracted_green_skill"])][
+                    ["full_skill_preferred_name", "extracted_full_skill_id"]
+                ]
+                .value_counts(normalize=True)[0:5]
+                .to_dict()
+            )
```
prop is nice

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/process_ojo_green_measures.py:
```diff
@@ -0,0 +1,753 @@
+                    }
+                )
+
+            top_5_sics_num = (
+                filtered_data[["SIC_2_digit", "SIC_2_digit_name"]]
+                .value_counts()[0:5]
+                .to_dict()
+            )
+            top_5_sics_prop = (
+                filtered_data[["SIC_2_digit", "SIC_2_digit_name"]]
+                .value_counts(normalize=True)[0:5]
+                .to_dict()
+            )
```
I can't decide if 2 digit is the way to go in the end... because it kind of obfuscates things - something to revisit!

india-kerle commented on dap_prinz_green_jobs/notebooks/README.md:
```diff
@@ -14,6 +14,10 @@ The `between_measures_analysis.ipynb` notebook contains analysis of green measur
 
```
there was a merge conflict with this file but i edited it very slightly to address !

india-kerle commented on dap_prinz_green_jobs/analysis/ojo_analysis/aggregate_by_region.py:
```diff
@@ -0,0 +1,93 @@
+
+    agg_itl_by = "itl_3_code"  # itl_2_code or itl_3_code
+
+    itl_aggregated_data = create_agg_data(
```
something i'd randomly be careful with/check is that London isn't always merged appropriately (a bug in the locations algorithm?) - I can add an issue to the `ojd_daps_locations` repo to investigate and mention to jack / might just want to have a cleaner in a utils somewhere

india-kerle reviewed (approved):
no major bugs that I flagged! I could run the scripts mentioned in the analysis README.md

I started going ham on asking for docstrings and typing hints but i think it might actually be overkill at this stage esp as the graphs are in live development - lets just save a day or two in the next sprint for clean up.

Something I did notice when I used locations data before is that for some unknown reason, not all of london is merged nicely all the time - might be worth poking around the location repo + adding a cleaner function eventually (not in this PR). 

Love the chloropleth!

lizgzil commented:
thanks so much @india-kerle ! really useful points. I added them all in this commit - https://github.com/nestauk/dap_prinz_green_jobs/pull/101/commits/6b4f00106e5377ad4dc0e9218ebe30dd0d61da32
I've added an analysis config which I think was a great shout!

india-kerle commented:
fabbo - Let me know when ya merge so i can use the helpful functions for the other graphs
